core/roboconf-target-iaas-ec2/src/main/java/net/roboconf/target/ec2/internal/Ec2MachineConfigurator.java

```diff
@@ -192,23 +235,131 @@ private boolean checkVmIsStarted() {
 		dis.setInstanceIds(Collections.singletonList(this.machineId));
 
 		DescribeInstancesResult disresult = this.ec2Api.describeInstances( dis );
+		// Obtain availability zone (for later use, eg. volume attachment).
+		this.availabilityZone = disresult.getReservations().get(0).getInstances().get(0).getPlacement().getAvailabilityZone();
 		return "running".equalsIgnoreCase( disresult.getReservations().get(0).getInstances().get(0).getState().getName());
 	}
+	
+	/**
+	 * Check whether EBS volume creation/attachment is requested.
+	 * @return true if requested, false otherwise
+	 */
+	private boolean volumeRequested() {
+		return Boolean.parseBoolean(this.targetProperties.get(Ec2Constants.USE_BLOCK_STORAGE));
+	}
+	
+	/**
+	 * Create volume for EBS.
+	 * @return volume ID if successful creation, null otherwise (or if nothing to do)
+	 */
+	private String createVolume(String snapshotId, int size) {
+		String volumeType = this.targetProperties.get(Ec2Constants.VOLUME_TYPE);
+		if(volumeType == null) volumeType = "standard";
+		
+		CreateVolumeRequest createVolumeRequest = new CreateVolumeRequest()
+		  .withAvailabilityZone(this.availabilityZone)
+		  .withVolumeType(volumeType)
+		  .withSize(size); // The size of the volume, in gigabytes.
+
+		// EC2 snapshot IDs start with "snap-"...
+		if(! Utils.isEmptyOrWhitespaces(snapshotId) && snapshotId.startsWith("snap-"))
+			createVolumeRequest.withSnapshotId(snapshotId);
+		
+		CreateVolumeResult createVolumeResult = this.ec2Api.createVolume(createVolumeRequest);
+		return createVolumeResult.getVolume().getVolumeId();
+	}
+	
+	/**
+	 * Check whether volume is created.
+	 * @param volumeId the EBS volume ID
+	 * @return true if volume created, false otherwise
+	 */
+	private boolean volumeCreated(String volumeId) {
+		DescribeVolumesRequest dvs = new DescribeVolumesRequest();
+        ArrayList<String> volumeIds = new ArrayList<String>();
+        volumeIds.add(volumeId);
+        dvs.setVolumeIds(volumeIds);
+        DescribeVolumesResult dvsresult = null;
+        try {
+        	dvsresult = this.ec2Api.describeVolumes(dvs);
+        } catch(Exception e) {
+        	dvsresult = null;
+        }
+        
+        return dvsresult != null && "available".equals(dvsresult.getVolumes().get(0).getState());
+	}
 
+	/**
+	 * Lookup volume, by ID or Name tag.
+	 * @param volumeIdOrName the EBS volume ID or Name tag
+	 * @return The volume ID of 1st matching volume found, null if no volume found
+	 */
+	private String lookupVolume(String volumeIdOrName) {
+		
+		if(! Utils.isEmptyOrWhitespaces(volumeIdOrName)) {
```

Comment:
Refactor the class to have a single return statement.
If we have to debug this function (in Eclipse or somewhere else), it will be much more easier.